File: solution/215KthLargestElementInAnArray/215KthLargestElementInAnArray.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
  int partition(vector<int> &nums, int lo, int hi) {
    int pilot_no = lo;
    int pilot = nums[lo];
    int k = lo + 1;
    while (k < hi) {
      if (nums[k] < pilot) {
        lo++;
        swap(nums[lo], nums[k]);
        k++;
      } else {
        hi--;
        swap(nums[k], nums[hi]);
      }
    }
    swap(nums[pilot_no], nums[lo]);
    return lo;
  }

public:
  int findKthLargest(vector<int> &nums, int k) {
    int lo = 0;
    int hi = nums.size();
    k = hi - k;
    while (lo < hi) {
      int pilot = partition(nums, lo, hi);
      if (pilot == k) {
        return nums[k];
      } else if (pilot < k) {
        lo = pilot + 1;
      } else {
        hi = pilot;
      }
    }
  }
};

int main() {
  Solution s;
  vector<int> nums1 = {3, 2, 1, 5, 6, 4};
  cout << s.findKthLargest(nums1, 4) << endl;
  vector<int> nums2 = {1, 1, 2, 3, 4, 5, 5, 5};
  cout << s.findKthLargest(nums2, 4) << endl;
  return 0;
}

File: src/215KthLargestElementInAnArray/215KthLargestElementInAnArray.rs
impl Solution {
    pub fn find_kth_largest(nums: Vec<i32>, k: i32) -> i32 {
        let mut nums = nums;
        let (mut lo, mut hi) = (0 as i32, nums.len() as i32 - 1);
        return Solution::find_kth_largest_helper(&mut nums, (hi+1-k) as usize, lo, hi);
    }
    
    fn find_kth_largest_helper(nums:&mut [i32], k: usize, mut lo: i32, mut hi: i32) -> i32 {
        let (mut pilot, mut i, mut j) = (lo, lo+1, hi);
        while i <= j {
            if nums[i as usize] <= nums[pilot as usize] {
                i += 1;
            }else{
                nums.swap(i as usize, j as usize);
                j -= 1;
            }
        }
        nums.swap(pilot as usize, j as usize);
        pilot = j;
        
        if pilot as usize == k {
            return nums[k];
        }else if (pilot as usize) < k{
            return Solution::find_kth_largest_helper(nums, k, pilot+1, hi);
        }else{
            return Solution::find_kth_largest_helper(nums, k, lo, pilot-1);
        }
    }
}
